File: agents/backend/onyx/server/features/robot_movement_ai/frontend/components/robot-3d-view/utils/compression.ts
/**
 * Data compression utilities
 * @module robot-3d-view/utils/compression
 */

/**
 * Compression options
 */
export interface CompressionOptions {
  level?: number; // 0-9, higher = better compression but slower
  method?: 'gzip' | 'deflate' | 'brotli';
}

/**
 * Compression Manager class
 */
export class CompressionManager {
  /**
   * Compresses data using CompressionStream API
   */
  async compress(data: string, options: CompressionOptions = {}): Promise<Uint8Array> {
    const { method = 'gzip' } = options;

    if (!('CompressionStream' in window)) {
      // Fallback: return original data as bytes
      return new TextEncoder().encode(data);
    }

    const stream = new CompressionStream(method);
    const writer = stream.writable.getWriter();
    const reader = stream.readable.getReader();

    // Write data
    writer.write(new TextEncoder().encode(data));
    writer.close();

    // Read compressed data
    const chunks: Uint8Array[] = [];
    let done = false;

    while (!done) {
      const { value, done: readerDone } = await reader.read();
      done = readerDone;
      if (value) {
        chunks.push(value);
      }
    }

    // Combine chunks
    const totalLength = chunks.reduce((acc, chunk) => acc + chunk.length, 0);
    const result = new Uint8Array(totalLength);
    let offset = 0;
    for (const chunk of chunks) {
      result.set(chunk, offset);
      offset += chunk.length;
    }

    return result;
  }

  /**
   * Decompresses data using DecompressionStream API
   */
  async decompress(data: Uint8Array, method: 'gzip' | 'deflate' | 'brotli' = 'gzip'): Promise<string> {
    if (!('DecompressionStream' in window)) {
      // Fallback: return data as string
      return new TextDecoder().decode(data);
    }

    const stream = new DecompressionStream(method);
    const writer = stream.writable.getWriter();
    const reader = stream.readable.getReader();

    // Write compressed data
    writer.write(data);
    writer.close();

    // Read decompressed data
    const chunks: Uint8Array[] = [];
    let done = false;

    while (!done) {
      const { value, done: readerDone } = await reader.read();
      done = readerDone;
      if (value) {
        chunks.push(value);
      }
    }

    // Combine chunks
    const totalLength = chunks.reduce((acc, chunk) => acc + chunk.length, 0);
    const result = new Uint8Array(totalLength);
    let offset = 0;
    for (const chunk of chunks) {
      result.set(chunk, offset);
      offset += chunk.length;
    }

    return new TextDecoder().decode(result);
  }

  /**
   * Compresses JSON data
   */
  async compressJSON(data: unknown): Promise<Uint8Array> {
    const json = JSON.stringify(data);
    return this.compress(json);
  }

  /**
   * Decompresses JSON data
   */
  async decompressJSON(data: Uint8Array): Promise<unknown> {
    const json = await this.decompress(data);
    return JSON.parse(json);
  }

  /**
   * Gets compression ratio
   */
  getCompressionRatio(original: number, compressed: number): number {
    return compressed / original;
  }

  /**
   * Checks if compression is supported
   */
  isCompressionSupported(): boolean {
    return 'CompressionStream' in window && 'DecompressionStream' in window;
  }
}

/**
 * Global compression manager instance
 */
export const compressionManager = new CompressionManager();

File: agents/backend/onyx/server/features/addiction_recovery_ai/utils/compression.py
"""
Compression utilities
Data compression functions
"""

from typing import Optional
import gzip
import zlib
import base64
import json


def compress_gzip(data: str) -> bytes:
    """
    Compress data using gzip
    
    Args:
        data: Data to compress
    
    Returns:
        Compressed data
    """
    return gzip.compress(data.encode('utf-8'))


def decompress_gzip(compressed: bytes) -> str:
    """
    Decompress gzip data
    
    Args:
        compressed: Compressed data
    
    Returns:
        Decompressed data
    """
    return gzip.decompress(compressed).decode('utf-8')


def compress_zlib(data: str) -> bytes:
    """
    Compress data using zlib
    
    Args:
        data: Data to compress
    
    Returns:
        Compressed data
    """
    return zlib.compress(data.encode('utf-8'))


def decompress_zlib(compressed: bytes) -> str:
    """
    Decompress zlib data
    
    Args:
        compressed: Compressed data
    
    Returns:
        Decompressed data
    """
    return zlib.decompress(compressed).decode('utf-8')


def compress_json(data: Any, method: str = "gzip") -> str:
    """
    Compress JSON data
    
    Args:
        data: Data to compress
        method: Compression method (gzip or zlib)
    
    Returns:
        Base64 encoded compressed data
    """
    json_str = json.dumps(data)
    
    if method == "gzip":
        compressed = compress_gzip(json_str)
    else:
        compressed = compress_zlib(json_str)
    
    return base64.b64encode(compressed).decode('utf-8')


def decompress_json(compressed: str, method: str = "gzip") -> Any:
    """
    Decompress JSON data
    
    Args:
        compressed: Base64 encoded compressed data
        method: Compression method (gzip or zlib)
    
    Returns:
        Decompressed data
    """
    compressed_bytes = base64.b64decode(compressed.encode('utf-8'))
    
    if method == "gzip":
        decompressed = decompress_gzip(compressed_bytes)
    else:
        decompressed = decompress_zlib(compressed_bytes)
    
    return json.loads(decompressed)


def compress_string(data: str, method: str = "gzip") -> str:
    """
    Compress string
    
    Args:
        data: String to compress
        method: Compression method (gzip or zlib)
    
    Returns:
        Base64 encoded compressed data
    """
    if method == "gzip":
        compressed = compress_gzip(data)
    else:
        compressed = compress_zlib(data)
    
    return base64.b64encode(compressed).decode('utf-8')


def decompress_string(compressed: str, method: str = "gzip") -> str:
    """
    Decompress string
    
    Args:
        compressed: Base64 encoded compressed data
        method: Compression method (gzip or zlib)
    
    Returns:
        Decompressed string
    """
    compressed_bytes = base64.b64decode(compressed.encode('utf-8'))
    
    if method == "gzip":
        return decompress_gzip(compressed_bytes)
    
    return decompress_zlib(compressed_bytes)


def get_compression_ratio(original: str, compressed: str) -> float:
    """
    Get compression ratio
    
    Args:
        original: Original data
        compressed: Compressed data (base64 encoded)
    
    Returns:
        Compression ratio
    """
    original_size = len(original.encode('utf-8'))
    compressed_size = len(base64.b64decode(compressed.encode('utf-8')))
    
    if original_size == 0:
        return 0.0
    
    return compressed_size / original_size

